src/dataflow/src/source/postgres.rs

```diff
@@ -197,26 +197,47 @@ impl PostgresSourceReader {
             .parse()
             .or_else(|_| Err(anyhow!("invalid lsn"))));
         for info in publication_tables {
-            // TODO(petrosagg): use a COPY statement here for more efficient network transfer
-            let data = client
-                .simple_query(&format!(
-                    "SELECT * FROM {:?}.{:?}",
-                    info.namespace, info.name
-                ))
+            let relation_id: Datum = (info.rel_id as i32).into();
+            let reader = client
+                .copy_out_simple(
+                    format!(
+                        "COPY {:?}.{:?} TO STDOUT (FORMAT TEXT, DELIMITER '\t')",
+                        info.namespace, info.name
+                    )
+                    .as_str(),
+                )
                 .await?;
-            for msg in data {
-                if let SimpleQueryMessage::Row(row) = msg {
+
+            pin_mut!(reader);
+            while let Some(Ok(b)) = reader.next().await {
+                // Convert raw rows from COPY into repr:Row. Each Row is a relation_id
+                // and list of string-encoded values, e.g. Row{ 16391 , ["1", "2"] }
+                let mut parser = pgcopy::CopyTextFormatParser::new(b.as_ref(), "\t", "\\N");
+                while !parser.is_eof() && !parser.is_end_of_copy_marker() {
                     let mut mz_row = Row::default();
-                    let rel_id: Datum = (info.rel_id as i32).into();
-                    mz_row.push(rel_id);
-                    mz_row.push_list((0..row.len()).map(|n| {
-                        let a: Datum = row.get(n).into();
-                        a
+                    mz_row.push(relation_id);
+
+                    try_fatal!(mz_row.push_list_with(|rp| -> Result<(), io::Error> {
+                        for col in 0..info.schema.len() {
+                            if col > 0 {
+                                parser.expect_column_delimiter()?;
```

Comment:
this looks like logic it should go in the `pgcopy` crate that you can call an `parser.iter_raw()` method and it should give you an iterator over raw values

Reply:
Ooo yes that'll be much cleaner